swap/swap.go

```diff
@@ -1,98 +1,636 @@
+type Owner struct {
+	Contract   common.Address    // address of swap contract
+	address    common.Address    // owner address
+	privateKey *ecdsa.PrivateKey // private key
+	publicKey  *ecdsa.PublicKey  // public key
+}
+
+// Params encapsulates param
+type Params struct {
+	InitialDepositAmount uint64
+}
+
+// NewParams returns a Params struct filled with default values
+func NewParams() *Params {
+	return &Params{
+		InitialDepositAmount: DefaultInitialDepositAmount,
+	}
 }
 
 // New - swap constructor
-func New(stateStore state.Store) (swap *Swap) {
-	swap = &Swap{
-		stateStore: stateStore,
-		balances:   make(map[enode.ID]int64),
+func New(stateStore state.Store, prvkey *ecdsa.PrivateKey, contract common.Address, backend contract.Backend) *Swap {
+	return &Swap{
+		store:               stateStore,
+		balances:            make(map[enode.ID]int64),
+		cheques:             make(map[enode.ID]*Cheque),
+		peers:               make(map[enode.ID]*Peer),
+		backend:             backend,
+		owner:               createOwner(prvkey, contract),
+		params:              NewParams(),
+		paymentThreshold:    DefaultPaymentThreshold,
+		disconnectThreshold: DefaultDisconnectThreshold,
+		oracle:              NewPriceOracle(),
 	}
-	return
 }
 
-//Swap implements the protocols.Balance interface
-//Add is the (sole) accounting function
+const (
+	balancePrefix        = "balance_"
+	sentChequePrefix     = "sent_cheque_"
+	receivedChequePrefix = "received_cheque_"
+)
+
+// returns the store key for retrieving a peer's balance
+func balanceKey(peer enode.ID) string {
+	return balancePrefix + peer.String()
+}
+
+// returns the store key for retrieving a peer's last sent cheque
+func sentChequeKey(peer enode.ID) string {
+	return sentChequePrefix + peer.String()
+}
+
+// returns the store key for retrieving a peer's last received cheque
+func receivedChequeKey(peer enode.ID) string {
+	return receivedChequePrefix + peer.String()
+}
+
+func keyToID(key string, prefix string) enode.ID {
+	return enode.HexID(key[len(prefix):])
+}
+
+// createOwner assings keys and addresses
+func createOwner(prvkey *ecdsa.PrivateKey, contract common.Address) *Owner {
+	pubkey := &prvkey.PublicKey
+	return &Owner{
+		privateKey: prvkey,
+		publicKey:  pubkey,
+		Contract:   contract,
+		address:    crypto.PubkeyToAddress(*pubkey),
+	}
+}
+
+// DeploySuccess is for convenience log output
+func (s *Swap) DeploySuccess() string {
+	return fmt.Sprintf("contract: %s, owner: %s, deposit: %v, signer: %x", s.owner.Contract.Hex(), s.owner.address.Hex(), s.params.InitialDepositAmount, s.owner.publicKey)
+}
+
+// Add is the (sole) accounting function
+// Swap implements the protocols.Balance interface
 func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+	s.accountingLock.Lock()
+	defer s.accountingLock.Unlock()
 
-	//load existing balances from the state store
-	err = s.loadState(peer)
+	err = s.loadBalance(peer.ID())
```

Comment:
@acud 's review suggested that we don't access the map directly, and that it would always be under a lock (which requires getter/setters a go maps are not safe). This has just happened in the last iteration of addressing reviews, which created now this new unfortunate situation which we didn't immediately realize, so first thanks for raising it. 

Maybe our first iteration wasn't that bad, in that the global `accountingLock` needs to be locking the whole `Add` run anyway, so accessing the maps directly wasn't that bad after all. Of course it makes the access to the maps dangerous *for the future*, but in the current state of affairs it was safe.

Another option would be to use go's slower and not type safe sync maps, which would not be that slow after all, considering that all accesses to `getBalance` now are with a dedicated lock, but it should be considered overkill as, again, in the current state of affairs everything is safe with the object's `accountingLock`.

In short, acknowledging that we may have misinterpreted previous reviews, I feel we got contradictory suggestions, so please advise on a best way forward.

Of course there is a third option, which is to redesign the complete flow, If that is the preferred way, please confirm.

Reply:
My opinion is to try to minimize the number of locks. There is probably an unnecessary locking for maps. This also can be optimized later.